### PR TITLE
Rnn base hidden size type check

### LINKED ISSUE #105631: torch.nn.RNNBase (and its subclasses) lacks parameter validation check
### Description:
When calling RNNBase or its subclasses(such as LSTM), set parameter 'hidden_size' to an illegal value will not be raised.

In detail, setting the parameter 'hidden_size' to 'True' (the incorrect type bool instead of int) does not cause the layer to reject the parameter in the checking statement. Instead, it is treated as 1 for normal computation and successfully passes the check condition ``proj_size(default 0) < hidden_size``. And all will be normal on cpu. However, when moving the parameter to CUDA, an error occurs due to the type not being int.

The performance is as below:

input (on cpu) :
```
p = torch.nn.LSTM(input_size=240, hidden_size=True)
x = torch.randn(1,15,240)
print(p(x)[0].shape)
```
output:
```
torch.Size([1, 15, 1])
```

input (on gpu) :
```
p = torch.nn.LSTM(input_size=240, hidden_size=True).to('cuda')
```
error message:
```
Traceback (most recent call last):
  File "D:\PythonProjects\venv\lib\site-packages\IPython\core\interactiveshell.py", line 3508, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-18-cb6fcc877e3f>", line 1, in <module>
    p = torch.nn.LSTM(input_size=240, hidden_size=True).to('cuda')
  File "D:\PythonProjects\venv\lib\site-packages\torch\nn\modules\module.py", line 1145, in to
    return self._apply(convert)
  File "D:\PythonProjects\venv\lib\site-packages\torch\nn\modules\rnn.py", line 202, in _apply
    self._init_flat_weights()
  File "D:\PythonProjects\venv\lib\site-packages\torch\nn\modules\rnn.py", line 139, in _init_flat_weights
    self.flatten_parameters()
  File "D:\PythonProjects\venv\lib\site-packages\torch\nn\modules\rnn.py", line 190, in flatten_parameters
    torch._cudnn_rnn_flatten_weight(
TypeError: _cudnn_rnn_flatten_weight(): argument 'hidden_size' (position 5) must be int, not bool

```

### Versions

PS D:\PythonProjects\venv\Lib\site-packages\torch\utils> python collect_env.py
Collecting environment information...
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 家庭中文版
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22621-SP0
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4070 Laptop GPU
Nvidia driver version: 532.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=2200
DeviceID=CPU0
Family=207
L2CacheSize=16384
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2200
Name=13th Gen Intel(R) Core(TM) i9-13900HX
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.1+cu117
[pip3] torchaudio==2.0.2+cu117
[pip3] torchvision==0.15.2+cu117
[conda] Could not collect

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @zou3519

### PR DESCRIPTION
**Pull request by franklinscudder**
Fixes #105631 

Added a type and value check on `hidden_size` to align behaviour between GPU and CPU modes and alert users when the wrong type is supplied.

### CODE REVIEW

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -72,6 +72,11 @@ def __init__(self, mode: str, input_size: int, hidden_size: int,
                           "recurrent layer, so non-zero dropout expects "
                           "num_layers greater than 1, but got dropout={} and "
                           "num_layers={}".format(dropout, num_layers))
+
+        if type(hidden_size) is not int:
```
Comment (albanD) [inline torch/nn/modules/rnn.py:76]:
```suggestion
        if isinstance(hidden_size, int):
```

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -72,6 +72,11 @@ def __init__(self, mode: str, input_size: int, hidden_size: int,
...
+            raise TypeError("hidden_size should be of type int, got: {}".format(type(hidden_size).__name__))
```
Comment (albanD) [inline torch/nn/modules/rnn.py:77]:
```suggestion
            raise TypeError(f"hidden_size should be of type int, got: {type(hidden_size).__name__)}"
```

**Review by albanD — COMMENTED** [review]
Small formatting nits, SGTM otherwise.

**Review by albanD — APPROVED** [review]
Thanks

File: torch/nn/modules/rnn.py [hunk 3]
```diff
@@ -72,6 +72,11 @@ def __init__(self, mode: str, input_size: int, hidden_size: int,
                           "recurrent layer, so non-zero dropout expects "
                           "num_layers greater than 1, but got dropout={} and "
                           "num_layers={}".format(dropout, num_layers))
+
+        if isinstance(hidden_size, int):
```
Comment (mikaylagawarecki) [inline torch/nn/modules/rnn.py:76]:
```suggestion
        if not isinstance(hidden_size, int):
```

**Comment by mikaylagawarecki** [comment]
Cancelling the merge as there is a small typo, please fix this! :)

File: torch/nn/modules/rnn.py (hunk 3 above)
Comment (franklinscudder) [reply torch/nn/modules/rnn.py]: Dammit! Nice catch :)

**Review by mikaylagawarecki — APPROVED** [review]

**Pull request closed without merging**